python/activator/middleware_interface.py

```diff
@@ -370,26 +371,55 @@ def __init__(self, central_butler: Butler, image_bucket: str, visit: FannedOutVi
         # How much to pad the spatial region we will copy over.
         self.padding = padding*lsst.geom.arcseconds
 
+    def _get_pp_hash(self):
+        """Get a unique ID for the Prompt Processing code.
+
+        Returns
+        ----------
+        hash : `hashlib.hash`
+            A hash of the contents of Prompt Processing.
+        """
+        root_dir = lsst.utils.getPackageDir("prompt_processing")
+        files = []
+        files.extend(sorted(glob.glob(os.path.join(root_dir, "maps", "**", "*.fits"), recursive=True)))
+        files.extend(sorted(glob.glob(os.path.join(root_dir, "pipelines", "**", "*.yaml"), recursive=True)))
+        # Source code has different paths in test environment and Docker container.
+        files.extend(sorted(glob.glob(os.path.join(root_dir, "python", "activator", "*.py"), recursive=True)))
+        files.extend(sorted(glob.glob(os.path.join(root_dir, "activator", "*.py"), recursive=True)))
+
+        filehash = hashlib.md5(usedforsecurity=False)
+        for file in files:
+            with open(file, "rb") as f:
+                filehash.update(f.read())  # Don't need to worry about OOM.
+        return filehash
+
     def _get_deployment(self):
         """Get a unique version ID of the active stack and pipeline configuration(s).
 
+        ``self._apdb_config`` must have already been initialized.
+
         Returns
         -------
         version : `str`
             A unique version identifier for the stack. Contains only
             word characters and "-", but not guaranteed to be human-readable.
         """
-        try:
-            # Defined by Knative in containers, guaranteed to be unique for
-            # each deployment. Currently of the form prompt-proto-service-#####.
-            return os.environ["K_REVISION"]
-        except KeyError:
-            # If not in a container, read the active Science Pipelines install.
-            packages = lsst.utils.packages.Packages.fromSystem()  # Takes several seconds!
-            h = hashlib.md5(usedforsecurity=False)
-            for package, version in packages.items():
-                h.update(bytes(package + version, encoding="utf-8"))
-            return f"local-{h.hexdigest()}"
+        # To disambiguate all stack changes, read the active Science Pipelines install.
+        packages = lsst.utils.packages.getAllPythonDistributions()
```

Comment:
Do you want to add a comment that this (by design) is not including non-python conda packages? (It will include conda python packages and pipeline python packages).

Reply:
I didn't realize there was such a thing as a "non-Python Conda package". 😵‍💫

Honestly, I'd classify that as "known issue" rather than "by design". Given how `rubin-env` is managed (i.e., lots of updates together), it shouldn't matter in practice.

Reply:
One of the great things about Conda is that it gives us an easy way to install a lot of libraries. If you look at the output of `conda list` you will see lots of `lib*` packages. Also `yaml` in conda-forge is the YAML C library.